active_discussion/embedded/critical-section.md

```diff
@@ -0,0 +1,290 @@
+  program.
+
+- I have observed that an `asm!("")` expression with *no* clobbers prevents
+  operations on `static mut` variables from being merged and reordered. See
+  example below. Is this intended behavior?
+
+``` rust
+#[no_mangle]
+static mut X: u32 = 0;
+
+#[no_mangle]
+unsafe fn INTERRUPT0() {
+    X += 1;
+
+    asm!("");
+
+    X += 2;
+}
+```
+
+Produces this machine code (sorry for the ARM assembly)
+
+``` armasm
+INTERRUPT0:
+        movw    r0, #0
+        movt    r0, #8192
+        ldr     r1, [r0]
+        adds    r1, #1
+        str     r1, [r0]  ; X += 1
+        ldr     r1, [r0]
+        adds    r1, #2
+        str     r1, [r0]  ; X += 2
+        bx      lr
+```
+
+This is the corresponding (post-optimization) LLVM IR:
+
+``` llvm
+; Function Attrs: nounwind
+define void @INTERRUPT0() unnamed_addr #1 !dbg !1268 {
+start:
+  %0 = load i32, i32* bitcast (<{ [4 x i8] }>* @X to i32*), align 4, !dbg !1269
+  %1 = add i32 %0, 1, !dbg !1269
+  store i32 %1, i32* bitcast (<{ [4 x i8] }>* @X to i32*), align 4, !dbg !1269
+  tail call void asm sideeffect "", ""() #5, !dbg !1270, !srcloc !1271
+  %2 = load i32, i32* bitcast (<{ [4 x i8] }>* @X to i32*), align 4, !dbg !1272
+  %3 = add i32 %2, 2, !dbg !1272
+  store i32 %3, i32* bitcast (<{ [4 x i8] }>* @X to i32*), align 4, !dbg !1272
+  ret void, !dbg !1273
+}
+```
+
+## Other comments
+
+(Feel free to disregard this section completely; it's about better optimizations
+rather than misoptimizations)
+
+`atomic::compiler_fence` and the "memory" clobber are coarse grained and they
+can prevent optimization of memory accesses that don't need to be synchronized.
+For example, this Rust code
+
+``` rust
+use core::{ptr, sync::atomic::{self, Ordering}};
+
+static mut X: u32 = 0;
+
+unsafe fn main() -> ! {
+    let mut y = 0;
+
+    // this could be part of a critical section
+    atomic::compiler_fence(Ordering::SeqCst);
+
+    y += 1;
+
+    // prevent the compiler from optimizing away `y`
+    unsafe {
+       ptr::read_volatile(&y);
+    }
+
+    loop {}
+}
+```
+
+produces this machine code
+
+``` armasm
+main:
+        sub     sp, #4
+        movs    r0, #0
+        str     r0, [sp]        ; y = 0
+        ldr     r0, [sp]
+        adds    r0, #1
+        str     r0, [sp]        ; y += 1
+        ldr     r0, [sp]        ; ptr::read_volatile
+        b       #-4 <main+0xe>
+```
+
+Without the compiler fence `y = 0` and `y += 1` would have been optimized into
+`y = 1` resulting in shorter machine code:
+
+``` armasm
+main:
+        sub     sp, #4
+        movs    r0, #1
+        str     r0, [sp]        ; y = 1
+        ldr     r0, [sp]        ; ptr::read_volatile
+        b       #-4 <main+0x8>
+```
+
+I wish we had a `atomic::compiler_fence_val(&x, ORDERING)` function that
+prevented only memory operations on `x` from being reordered -- though I don't
+know if LLVM supports that kind of fine grained compiler fences -- that would
+result in better optimizations.
```

Comment:
If `x` where to be a pointer here, that would not prevent re-orderings on memory that can be reached from `x`, only on `x` itself. Would that be ok?